From 0818150f6dd45500373a7b746a89a8dd75a7ae4c Mon Sep 17 00:00:00 2001
From: Sergiy Kibrik <sakib@meta.ua>
Date: Sat, 5 Oct 2013 07:58:51 +0300
Subject: [PATCH 1/2] add scalable GUI theme

Add and switch to gm-scalable, which places and scales UI elements
according to screen width/height. It makes theme applicable to
different screen sizes, e.g. 600x800 (most e-readers' displays),
800x600 (same, only horizontal layout), 640x480 (QEMU), but also
should be robust to any resolution/ratio.
---
 src/gm.h                       |    2 +-
 themes/default/gm-scalable.edc |  843 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 844 insertions(+), 1 deletion(-)
 create mode 100644 themes/default/gm-scalable.edc

diff --git a/src/gm.h b/src/gm.h
index d67d0c8..4c2d164 100644
--- a/src/gm.h
+++ b/src/gm.h
@@ -3,7 +3,7 @@
 
 #include <libkeys.h>
 
-#define THEME_EDJE "gm"
+#define THEME_EDJE "gm-scalable"
 
 keys_t* gm_keys();
 
diff --git a/themes/default/gm-scalable.edc b/themes/default/gm-scalable.edc
new file mode 100644
index 0000000..cca4738
--- /dev/null
+++ b/themes/default/gm-scalable.edc
@@ -0,0 +1,843 @@
+#define CONCAT(x, y)  x##y
+#define TRIPLER(x, y, z) x##y##z
+#define STR(x) XSTR(x)
+#define XSTR(x) #x
+
+images {
+    image, "start-menu-1-left-bottom-corner.png" COMP;
+    image, "start-menu-1-right-bottom-corner.png" COMP;
+    image, "start-menu-date-left-bottom-corner.png" COMP;
+    image, "start-menu-date-right-bottom-corner.png" COMP;
+
+    image, "start-menu-left-top-corner-small.png" COMP;
+    image, "start-menu-right-top-corner-small.png" COMP;
+    image, "start-menu-left-bottom-corner-small.png" COMP;
+    image, "start-menu-right-bottom-corner-small.png" COMP;
+    image, "start-menu-binder-small.png" COMP;
+}
+
+styles {
+    style {
+        name, "caption";
+        base, "font=sans-serif font_size=20 color=#000 align=middle";
+        tag, "inactive" "+color=#bbb";
+        tag, "/inactive" "-";
+    }
+    style {
+        name, "day";
+        base, "font=sans-serif font_size=60 color=#000 align=middle";
+    }
+    style {
+        name, "clock";
+        base, "font=sans-serif font_size=40 color=#000 align=middle";
+    }
+}
+
+#define COVER_IMAGE(x1, y1, x2, y2)                              \
+    part {                                                       \
+        name, "cover_image";                                     \
+        type, SWALLOW;                                           \
+        description {                                            \
+            state, "default" 0.0;                                \
+            rel1 { relative, x1 y1; }                            \
+            rel2 { relative, x2 y2; }                            \
+        }                                                        \
+    }
+
+#define ITEM(N, NAME, X1, Y1, X2, Y2)                                   \
+            part {                                                      \
+                name, STR(N-bg);                                        \
+                type, RECT;                                             \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    color, 210 210 210 255;                             \
+                    rel1 { relative, X1 Y1; }                           \
+                    rel2 { relative, X2 Y2; }                           \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-top-bg);                                    \
+                type, RECT;                                             \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    color, 85 85 85 255;                                \
+                    rel1 { to, STR(N-bg); offset, 0 -10; }              \
+                    rel2 { to, STR(N-bg); relative, 1.0 0.0; offset, -1 -2; } \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-top);                                       \
+                images { image, STR(start-menu-label-N.png) COMP; }     \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-top-bg); relative, 0.42 0.0; offset, 0 -15; } \
+                    rel2 { to, STR(N-top-bg); relative, 0.58 1.0; offset, 0 -1; } \
+                    image { normal, STR(start-menu-label-N.png); }      \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-top-left);                                  \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-top-bg); }                         \
+                    rel2 { to, STR(N-top-bg); relative, 0.0 1.0; offset, 3 -1; } \
+                    image { normal, "start-menu-left-top-corner-small.png"; } \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-top-right);                                 \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-top-bg); relative, 1.0 0.0; offset, -4 0; } \
+                    rel2 { to, STR(N-top-bg); }                        \
+                    image { normal, "start-menu-right-top-corner-small.png"; } \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-binder);                                    \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-bg); offset, 0 5; }                \
+                    rel2 { to, STR(N-bg); relative, 1.0 0.0; offset, -1 25; } \
+                    image { normal, "start-menu-binder-small.png"; }    \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-icon);                                      \
+                images { image, STR(start-menu-icon-NAME.png) COMP; }   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-bg); offset, 31 39; }              \
+                    rel2 { to, STR(N-bg); offset, -32 -40; }            \
+                    image { normal, STR(start-menu-icon-NAME.png); }    \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(CONCAT(caption_,NAME));                       \
+                type, TEXTBLOCK;                                        \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    color, 0 0 0 255;                                   \
+                    text { style, "caption"; }                          \
+                    align, 0.5 0.0;                                     \
+                    rel1 { to, STR(N-bg); relative, 0.0 0.76; }         \
+                    rel2 { to, STR(N-bg); offset, -1 -1; }              \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-bottom-left);                               \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-bg); relative, 0.0 1.0; offset, 0 -2; } \
+                    rel2 { to, STR(N-bg); relative, 0.0 1.0; offset, 1 -1; } \
+                    image { normal, "start-menu-left-bottom-corner-small.png"; } \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, STR(N-bottom-right);                              \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, STR(N-bg); relative, 1.0 1.0; offset, -2 -2; } \
+                    rel2 { to, STR(N-bg); relative, 1.0 1.0; offset, -1 -1; } \
+                    image { normal, "start-menu-right-bottom-corner-small.png"; } \
+                }                                                       \
+            };                                                          \
+
+
+#define CURSOR_STATE(NAME, X1, Y1, X2, Y2, C)   \
+            description {                       \
+                state, STR(NAME) 0.0;           \
+                inherit, "default" 0.0;         \
+                color, C C C 255;               \
+                rel1 {                          \
+                    relative, X1 Y1;            \
+                }                               \
+                rel2 {                          \
+                    relative, X2 Y2;            \
+                }                               \
+            }
+
+#define CURSOR_PROGRAM(name_)                            \
+    program {                                            \
+        signal, "cursor_set";                            \
+        source, STR(name_);                              \
+        action, STATE_SET STR(name_) 0.0;                \
+        target, "cursor-bg";                             \
+    }
+
+#define CURSOR_SELECT_PROGRAM                                   \
+            /* Figure out what item to run and emit it back */  \
+            program {                                           \
+                signal, "select";                               \
+                script {                                        \
+                    new st[256];                                \
+                    new Float:f;                                \
+                    get_state(PART:"cursor-bg", st, 255, f);    \
+                    emit("activate", st);                       \
+                }                                               \
+            }
+
+#define CURSOR_SET_PROGRAM                      \
+            /* Move cursor to specified item */ \
+            program {                           \
+                signal, "cursor_set";           \
+                source, "default";              \
+                action, STATE_SET "Books" 0.0;  \
+                target, "cursor-bg";            \
+            }
+
+#define CURSOR_MOVE(DIR)                                            \
+    program {                                                       \
+        signal, "move";                                             \
+        source, STR(DIR);                                           \
+        script {                                                    \
+            new st[256];                                            \
+            new Float:f;                                            \
+            get_state(PART:"cursor-bg", st, 255, f);
+
+#define END_CURSOR_MOVE() {}}}
+
+#define _MOVE(FROM, TO)                            \
+    if (!strcmp(st, STR(FROM))) {                  \
+        set_state(PART:"cursor-bg", STR(TO), 0.0); \
+    } else
+
+#define CURSOR_GRAPHICS                                                 \
+            part {                                                      \
+                name, "cursor-left-top";                                \
+                images { image, "cursor-frame-left-top.png" COMP; }     \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; }                           \
+                    rel2 { to, "cursor-bg"; relative, 0.0 0.0; offset, 7 7; } \
+                    image { normal, "cursor-frame-left-top.png"; }      \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-top";                                     \
+                type, RECT;                                             \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; offset, 8 0; }              \
+                    rel2 { to, "cursor-bg"; relative, 1.0 0.0; offset, -9 2; } \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-right-top";                               \
+                images { image, "cursor-frame-right-top.png" COMP; }    \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; relative, 1.0 0.0; offset, -8 0; } \
+                    rel2 { to, "cursor-bg"; relative, 1.0 0.0; offset, -1 7; } \
+                    image { normal, "cursor-frame-right-top.png"; }     \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-right";                                   \
+                type, RECT;                                             \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; relative, 1.0 0.0; offset, -3 8; } \
+                    rel2 { to, "cursor-bg"; offset, -1 -9; }            \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-right-bottom";                            \
+                images { image, "cursor-frame-right-bottom.png" COMP; } \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; relative, 1.0 1.0; offset, -8 -8; } \
+                    rel2 { to, "cursor-bg"; }                           \
+                    image { normal, "cursor-frame-right-bottom.png"; }  \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-bottom";                                  \
+                type, RECT;                                             \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; relative, 0.0 1.0; offset, 8 -3; } \
+                    rel2 { to, "cursor-bg"; offset, -9 -1; }            \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-left-bottom";                             \
+                images { image, "cursor-frame-left-bottom.png" COMP; }  \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; relative, 0.0 1.0; offset, 0 -8; } \
+                    rel2 { to, "cursor-bg"; relative, 0.0 1.0; offset, 7 -1; } \
+                    image { normal, "cursor-frame-left-bottom.png"; }   \
+                }                                                       \
+            }                                                           \
+            part {                                                      \
+                name, "cursor-left";                                    \
+                type, RECT;                                             \
+                clip_to, "cursor-bg";                                   \
+                description {                                           \
+                    state, "default" 0.0;                               \
+                    rel1 { to, "cursor-bg"; offset, 0 8; }              \
+                    rel2 { to, "cursor-bg"; relative, 0.0 1.0; offset, 2 -9; } \
+                }                                                       \
+            }
+
+#define BOOK_GRAPHICS 							\
+            part {							\
+                name, "1-top";						\
+                description {						\
+                    state, "default" 0.0;				\
+                    rel1 { to, "1-bg"; offset, 0 -24; }			\
+                    rel2 { to, "1-bg"; relative, 1.0 0.0; offset, -1 -1; } \
+                    image { normal, "start-menu-1-top.png"; }		\
+                }							\
+            }								\
+            part {							\
+                name, "1-splitter";					\
+                description {						\
+                    state, "default" 0.0;				\
+                    rel1 { to, "1-bg"; relative, 0.0 0.72; }		\
+                    rel2 { to, "1-bg"; relative, 1.0 0.72; offset, -1 2; } \
+                    image { normal, "start-menu-1-splitter.png"; }	\
+                }							\
+            }								\
+            part {							\
+                name, "1-left-bottom-corner";				\
+                description {						\
+                    state, "default" 0.0;				\
+                    rel1 { to, "1-bg"; relative, 0.0 1.0; offset, 0 -2; } \
+                    rel2 { to, "1-bg"; relative, 0.0 1.0; offset, 1 -1; } \
+                    image { normal, "start-menu-1-left-bottom-corner.png"; } \
+                }							\
+            }								\
+            part {							\
+                name, "1-right-bottom-corner";				\
+                description {						\
+                    state, "default" 0.0;				\
+                    rel1 { to, "1-bg"; relative, 1.0 1.0; offset, -2 -2; } \
+                    rel2 { to, "1-bg"; relative, 1.0 1.0; offset, -1 -1; } \
+                    image { normal, "start-menu-1-right-bottom-corner.png"; } \
+                }							\
+            }
+
+#define DATE_GRAPHICS \
+            part {							\
+                name, "date-top";					\
+                description {						\
+                    state, "default" 0.0;				\
+                    rel1 {						\
+                        to, "date-bg";					\
+                        offset, 0 -37;					\
+                    }							\
+                    rel2 {						\
+                        to, "date-bg";					\
+                        relative, 1.0 0.0;				\
+                        offset, -1 -1;					\
+                    }							\
+                    image {						\
+                        normal, "start-menu-date-top.png";		\
+                    }							\
+                }							\
+            }								\
+            part {							\
+                name, "date-left-bottom-corner";			\
+                description {						\
+                    state, "default" 0.00;				\
+                    rel1 { to, "date-bg"; relative, 0.0 1.0; offset, 0 -2; } \
+                    rel2 { to, "date-bg"; relative, 0.0 1.0; offset, 0 -1; } \
+                    image { normal, "start-menu-date-left-bottom-corner.png"; } \
+                }							\
+            }								\
+            part {							\
+                name, "date-right-bottom-corner";			\
+                description {						\
+                    state, "default" 0.0;				\
+                    rel1 { to, "date-bg"; relative, 1.0 1.0; offset, -1 -2; } \
+                    rel2 { to, "date-bg"; relative, 1.0 1.0; offset, -1 -1; } \
+                    image { normal, "start-menu-date-right-bottom-corner.png"; } \
+                }							\
+            }
+
+
+collections {
+    group {
+        name, "text-menu";
+        data {
+            item, "title" "";
+        }
+    }
+
+    group {
+        name, "vertical_graphics";
+        images {
+            image, "start-menu-1-top.png" COMP;
+            image, "start-menu-1-splitter.png" COMP;
+            image, "start-menu-date-top.png" COMP;
+        }
+        parts {
+            part {
+                name, "bg";
+                type, RECT;
+                description {
+                    state, "default" 0.0;
+                    color, 255 255 255 255;
+                }
+            }
+
+#define ROW0_Y1  0.05
+#define ROW0_Y2  0.45
+#define ROW1_Y1  0.495
+#define ROW1_Y2  0.72
+#define ROW2_Y1  0.76
+#define ROW2_Y2  0.983
+
+#define COL1_X1  0.03
+#define COL1_X2  0.323
+#define COL2_X1  0.35
+#define COL2_X2  0.646
+#define COL3_X1  0.673
+#define COL3_X2  0.97
+
+            /* -- Book -- */
+            part {
+                name, "1-bg";
+                type, RECT;
+                description {
+                    state, "default" 0.0;
+                    color, 85 85 85 255;
+                    rel1 { relative, COL1_X1 ROW0_Y1; }
+                    rel2 { relative, 0.47 ROW0_Y2; }
+                }
+            }
+
+            BOOK_GRAPHICS;
+
+            /* -- Date -- */
+            part {
+                name, "date-bg";
+                type, RECT;
+                description {
+                    state, "default" 0.0;
+                    color, 210 210 210 255;
+                    rel1 { relative, 0.5 ROW0_Y1; offset, 0 28; }
+                    rel2 { relative, COL3_X2 ROW0_Y2; }
+                }
+            }
+
+            DATE_GRAPHICS;
+
+            /* -- 2-7 -- */
+
+            ITEM(2, library, COL1_X1, ROW1_Y1, COL1_X2, ROW1_Y2);
+            ITEM(3, photo,   COL2_X1, ROW1_Y1, COL2_X2, ROW1_Y2);
+            ITEM(4, audio,   COL3_X1, ROW1_Y1, COL3_X2, ROW1_Y2);
+            ITEM(5, games,   COL1_X1, ROW2_Y1, COL1_X2, ROW2_Y2);
+            ITEM(6, apps,    COL2_X1, ROW2_Y1, COL2_X2, ROW2_Y2);
+            ITEM(7, setup,   COL3_X1, ROW2_Y1, COL3_X2, ROW2_Y2);
+
+            /* -- Cursor -- */
+
+#define CURSOR_ROW1_Y1  0.5
+#define CURSOR_ROW1_Y2  ROW1_Y2
+#define CURSOR_ROW2_Y1  0.8
+#define CURSOR_ROW2_Y2  ROW2_Y2
+
+            part {
+                name, "cursor-bg";
+                type, RECT;
+
+                description { state, "default" 0.0; }
+                CURSOR_STATE(Book,   COL1_X1, 0.052,          0.47,    ROW0_Y2, 210);
+                CURSOR_STATE(Date,   0.5,     0.082,          COL3_X2, ROW0_Y2, 85);
+                CURSOR_STATE(Books,  COL1_X1, CURSOR_ROW1_Y1, COL1_X2, CURSOR_ROW1_Y2, 85);
+                CURSOR_STATE(Images, COL2_X1, CURSOR_ROW1_Y1, COL2_X2, CURSOR_ROW1_Y2, 85);
+                CURSOR_STATE(Audio,  COL3_X1, CURSOR_ROW1_Y1, COL3_X2, CURSOR_ROW1_Y2, 85);
+                CURSOR_STATE(Games,  COL1_X1, CURSOR_ROW2_Y1, COL1_X2, CURSOR_ROW2_Y2, 85);
+                CURSOR_STATE(Apps,   COL2_X1, CURSOR_ROW2_Y1, COL2_X2, CURSOR_ROW2_Y2, 85);
+                CURSOR_STATE(Setup,  COL3_X1, CURSOR_ROW2_Y1, COL3_X2, CURSOR_ROW2_Y2, 85);
+            }
+
+            CURSOR_GRAPHICS
+
+#define DAY_X1   0.5
+#define DAY_Y1   0.145
+#define DAY_X2   0.97
+#define DAY_Y2   0.26
+
+#define CAPTION(t, x1, y1, x2, y2) \
+            part {				\
+                name, t;			\
+                type, TEXTBLOCK;		\
+                mouse_events, 0;		\
+                description {			\
+                    state, "default" 0.0;	\
+                    color, 0 0 0 255;		\
+                    text {			\
+                        style, "caption";	\
+                        max, 0 1;		\
+                    }				\
+                    align, 0.5 0.5;		\
+                    rel1 { relative, x1 y1; }	\
+                    rel2 { relative, x2 y2; }	\
+                }				\
+            }
+
+
+            CAPTION("caption_month", DAY_X1, 0.09, DAY_X2, 0.14)
+
+            COVER_IMAGE(COL1_X1, ROW0_Y1, 0.458, 0.331);
+            part {
+                name, "caption_title";
+                type, TEXT;
+                description {
+                    state, "default" 0.0;
+                    color, 255 255 255 255;
+                    text {
+                        text, "";
+                        font, "sans-serif";
+                        size, 24;
+                        align, 0.0 0.5;
+                        elipsis, 0.0;
+                    }
+                    rel1 { relative, 0.04 0.35; }
+                    rel2 { relative, 0.46 0.38; }
+                }
+            }
+
+            part {
+                name, "caption_author";
+                type, TEXT;
+                description {
+                    state, "default" 0.0;
+                    color, 255 255 255 255;
+                    text {
+                        text, "";
+                        font, "sans-serif";
+                        size, 24;
+                        align, 0.0 0.5;
+                        elipsis, 0.0;
+                    }
+                    rel1 { relative, 0.04 0.39; }
+                    rel2 { relative, 0.46 0.44; }
+                }
+            }
+
+#define STYLED(t, s, x1, y1, x2, y2)		\
+            part {				\
+                name, t;			\
+                type, TEXTBLOCK;		\
+                mouse_events, 0;		\
+                description {			\
+                    state, "default" 0.0;	\
+                    color, 0 0 0 255;		\
+                    text {			\
+                        style, s;		\
+                        max, 0 1;		\
+                    }				\
+                    align, 0.5 0.5;		\
+                    rel1 { relative, x1 y1; }	\
+                    rel2 { relative, x2 y2; }	\
+                }				\
+            }
+
+            STYLED("caption_day", "day", DAY_X1, DAY_Y1, DAY_X2, DAY_Y2)
+            STYLED("caption_clock", "clock", 0.56, 0.34, 0.89, 0.43)
+            CAPTION("caption_dayofweek", DAY_X1, 0.26, DAY_X2, 0.04)
+        }
+
+        programs {
+            CURSOR_SELECT_PROGRAM;
+            CURSOR_SET_PROGRAM;
+            CURSOR_PROGRAM(Book);
+            CURSOR_PROGRAM(Date);
+            CURSOR_PROGRAM(Books);
+            CURSOR_PROGRAM(Images);
+            CURSOR_PROGRAM(Audio);
+            CURSOR_PROGRAM(Games);
+            CURSOR_PROGRAM(Apps);
+            CURSOR_PROGRAM(Setup);
+
+            /* Cursor movement handing */
+            CURSOR_MOVE(CursorLeft)
+                _MOVE(Book, Setup)
+                _MOVE(Date, Book)
+                _MOVE(Books, Date)
+                _MOVE(Images, Books)
+                _MOVE(Audio, Images)
+                _MOVE(Games, Audio)
+                _MOVE(Apps, Games)
+                _MOVE(Setup, Apps)
+            END_CURSOR_MOVE();
+
+            CURSOR_MOVE(CursorUp)
+                _MOVE(Book, Games)
+                _MOVE(Date, Setup)
+                _MOVE(Books, Book)
+                _MOVE(Images, Book)
+                _MOVE(Audio, Date)
+                _MOVE(Games, Books)
+                _MOVE(Apps, Images)
+                _MOVE(Setup, Audio)
+            END_CURSOR_MOVE();
+
+            CURSOR_MOVE(CursorRight)
+                _MOVE(Book, Date)
+                _MOVE(Date, Books)
+                _MOVE(Books, Images)
+                _MOVE(Images, Audio)
+                _MOVE(Audio, Games)
+                _MOVE(Games, Apps)
+                _MOVE(Apps, Setup)
+                _MOVE(Setup, Book)
+            END_CURSOR_MOVE();
+
+            CURSOR_MOVE(CursorDown)
+                _MOVE(Book, Books)
+                _MOVE(Date, Audio)
+                _MOVE(Books, Games)
+                _MOVE(Images, Apps)
+                _MOVE(Audio, Setup)
+                _MOVE(Games, Book)
+                _MOVE(Apps, Book)
+                _MOVE(Setup, Date)
+            END_CURSOR_MOVE();
+        }
+    }
+
+    group {
+        name, "horizontal_graphics";
+        parts {
+            part {
+                name, "bg";
+                type, RECT;
+                description {
+                    state, "default" 0.0;
+                    color, 255 255 255 255;
+                }
+            }
+
+#undef ROW0_Y1
+#undef ROW0_Y2
+#undef ROW1_Y1
+#undef ROW1_Y2
+#undef ROW2_Y1
+#undef ROW2_Y2
+
+#undef COL1_X1
+#undef COL1_X2
+#undef COL2_X1
+#undef COL2_X2
+#undef COL3_X1
+#undef COL3_X2
+#undef COL4_X1
+#undef COL4_X2
+
+#undef CURSOR_ROW1_Y1
+#undef CURSOR_ROW1_Y2
+#undef CURSOR_ROW2_Y1
+#undef CURSOR_ROW2_Y2
+#undef CURSOR_ROW3_Y1
+#undef CURSOR_ROW3_Y2
+
+#define ROW0_Y1  0.07
+#define ROW0_Y2  0.26
+#define ROW1_Y1  0.31
+#define ROW1_Y2  0.61
+#define ROW2_Y1  0.67
+#define ROW2_Y2  0.97
+
+#define COL1_X1  0.02
+#define COL1_X2  0.24
+#define COL2_X1  0.26
+#define COL2_X2  0.48
+#define COL3_X1  0.51
+#define COL3_X2  0.73
+#define COL4_X1  0.75
+#define COL4_X2  0.98
+
+            /* -- Book -- */
+
+            part {
+                name, "1-bg";
+                type, RECT;
+                description {
+                    state, "default" 0.0;
+                    color, 85 85 85 255;
+                    rel1 { relative, COL1_X1 ROW0_Y1; }
+                    rel2 { relative, COL2_X2 ROW1_Y2; }
+                }
+            }
+
+            BOOK_GRAPHICS;
+
+            /* -- Date -- */
+            part {
+                name, "date-bg";
+                type, RECT;
+                description {
+                    state, "default" 0.0;
+                    color, 210 210 210 255;
+                    rel1 { relative, COL3_X1 ROW0_Y1; offset, 0 28; }
+                    rel2 { relative, COL4_X2 ROW0_Y2; }
+                }
+            }
+
+            DATE_GRAPHICS;
+
+            /* -- 2-7 -- */
+
+            ITEM(2, library, COL3_X1, ROW1_Y1, COL3_X2, ROW1_Y2);
+            ITEM(3, photo,   COL4_X1, ROW1_Y1, COL4_X2, ROW1_Y2);
+            ITEM(4, audio,   COL1_X1, ROW2_Y1, COL1_X2, ROW2_Y2);
+            ITEM(5, games,   COL2_X1, ROW2_Y1, COL2_X2, ROW2_Y2);
+            ITEM(6, apps,    COL3_X1, ROW2_Y1, COL3_X2, ROW2_Y2);
+            ITEM(7, setup,   COL4_X1, ROW2_Y1, COL4_X2, ROW2_Y2);
+
+            /* -- Cursor -- */
+
+#define CURSOR_ROW1_Y1  0.36
+#define CURSOR_ROW1_Y2  0.61
+#define CURSOR_ROW2_Y1  0.72
+#define CURSOR_ROW2_Y2  0.97
+
+            part {
+                name, "cursor-bg";
+                type, RECT;
+
+                description { state, "default" 0.0; }
+                CURSOR_STATE(Book,   COL1_X1, ROW0_Y1,        COL2_X2, ROW1_Y2, 210);
+                CURSOR_STATE(Date,   COL3_X1, 0.12,           COL4_X2, ROW0_Y2, 85);
+                CURSOR_STATE(Books,  COL3_X1, CURSOR_ROW1_Y1, COL3_X2, CURSOR_ROW1_Y2, 85);
+                CURSOR_STATE(Images, COL4_X1, CURSOR_ROW1_Y1, COL4_X2, CURSOR_ROW1_Y2, 85);
+                CURSOR_STATE(Audio,  COL1_X1, CURSOR_ROW2_Y1, COL1_X2, CURSOR_ROW2_Y2, 85);
+                CURSOR_STATE(Games,  COL2_X1, CURSOR_ROW2_Y1, COL2_X2, CURSOR_ROW2_Y2, 85);
+                CURSOR_STATE(Apps,   COL3_X1, CURSOR_ROW2_Y1, COL3_X2, CURSOR_ROW2_Y2, 85);
+                CURSOR_STATE(Setup,  COL4_X1, CURSOR_ROW2_Y1, COL4_X2, CURSOR_ROW2_Y2, 85);
+            }
+
+            CURSOR_GRAPHICS;
+
+            COVER_IMAGE(0.02, 0.07, 0.47, 0.44);
+            part {
+                name, "caption_title";
+                type, TEXT;
+                description {
+                    state, "default" 0.0;
+                    text {
+                        text, "";
+                        font, "sans-serif";
+                        size, 24;
+                        align, 0.0 0.5;
+                        elipsis, 0.0;
+                    }
+                    rel1 { relative, 0.03 0.46; }
+                    rel2 { relative, 0.47 0.51; }
+                }
+            }
+
+            part {
+                name, "caption_author";
+                type, TEXT;
+                description {
+                    state, "default" 0.0;
+                    text {
+                        text, "";
+                        font, "sans-serif";
+                        size, 24;
+                        align, 0.0 0.5;
+                        elipsis, 0.0;
+                    }
+                    rel1 { relative, 0.03 0.52; }
+                    rel2 { relative, 0.47 0.59; }
+                }
+            }
+
+            STYLED("caption_clock", "clock", 0.75, 0.11, 0.97, 0.25)
+            part {
+                name, "caption_date";
+                type, TEXTBLOCK;
+                description {
+                    state, "default" 0.0;
+                    text { style, "caption"; }
+                    rel1 { to, "date-bg"; relative, 0.1 0.05; }
+                    rel2 { to, "date-bg"; relative, 0.5 0.45; }
+                }
+            }
+
+            part {
+                name, "caption_dayofweek";
+                type, TEXTBLOCK;
+                description {
+                    state, "default" 0.0;
+                    text { style, "caption"; max, 0 1; }
+                    align, 0.0 0.9;
+                    rel1 { to, "date-bg"; relative, 0.1 0.55; }
+                    rel2 { to, "date-bg"; relative, 0.5 0.95; }
+                }
+            }
+        }
+        programs {
+            CURSOR_SELECT_PROGRAM;
+            CURSOR_SET_PROGRAM;
+            CURSOR_PROGRAM(Book);
+            CURSOR_PROGRAM(Date);
+            CURSOR_PROGRAM(Books);
+            CURSOR_PROGRAM(Images);
+            CURSOR_PROGRAM(Audio);
+            CURSOR_PROGRAM(Games);
+            CURSOR_PROGRAM(Apps);
+            CURSOR_PROGRAM(Setup);
+
+            /* Cursor movement handing */
+            CURSOR_MOVE(CursorLeft)
+                _MOVE(Book, Setup)
+                _MOVE(Date, Book)
+                _MOVE(Books, Book)
+                _MOVE(Images, Books)
+                _MOVE(Audio, Images)
+                _MOVE(Games, Audio)
+                _MOVE(Apps, Games)
+                _MOVE(Setup, Apps)
+            END_CURSOR_MOVE();
+
+            CURSOR_MOVE(CursorUp)
+                _MOVE(Book, Audio)
+                _MOVE(Date, Setup)
+                _MOVE(Books, Date)
+                _MOVE(Images, Date)
+                _MOVE(Audio, Book)
+                _MOVE(Games, Book)
+                _MOVE(Apps, Books)
+                _MOVE(Setup, Images)
+            END_CURSOR_MOVE();
+
+            CURSOR_MOVE(CursorRight)
+                _MOVE(Book, Books)
+                _MOVE(Date, Book)
+                _MOVE(Books, Images)
+                _MOVE(Images, Audio)
+                _MOVE(Audio, Games)
+                _MOVE(Games, Apps)
+                _MOVE(Apps, Setup)
+                _MOVE(Setup, Book)
+            END_CURSOR_MOVE();
+
+            CURSOR_MOVE(CursorDown)
+                _MOVE(Book, Audio)
+                _MOVE(Date, Books)
+                _MOVE(Books, Apps)
+                _MOVE(Images, Setup)
+                _MOVE(Audio, Book)
+                _MOVE(Games, Book)
+                _MOVE(Apps, Date)
+                _MOVE(Setup, Date)
+            END_CURSOR_MOVE();
+        }
+    }
+}
-- 
1.7.9.5

